File: dsaWithJava/searching/BinarySearch.java
package searching;

public class BinarySearch {
    public static int binarysearchIterative(int[] arr, int num) {
        return binarysearchIterative(arr, num, 0, arr.length - 1);
    }

    private static int binarysearchIterative(int[] arr, int num, int begin, int end) {
        while (begin <= end) {
            int mid = (begin + end) / 2;
            if (arr[mid] < num)
                begin = mid + 1;
            else if (arr[mid] > num)
                end = mid - 1;
            else
                return mid;
        }
        return -1;
    }

    public static int binarySearchRecursive(int[] arr, int num) {
        return binarySearchRecursive(arr, num, 0, arr.length - 1);
    }

    private static int binarySearchRecursive(int[] arr, int num, int begin, int end) {
        if (begin > end)
            return -1;
        int mid = (begin + end) / 2;
        if (arr[mid] == num)
            return mid;
        if (arr[mid] > num)
            return binarySearchRecursive(arr, num, begin, mid - 1);
        else
            return binarySearchRecursive(arr, num, mid + 1, end);
    }
}

File: dsa_with_python/searching/binary_search.py
def binary_search_iterative(arr: list, num: int) -> int:
    return __binary_search_iterate(arr, num, 0, len(arr) - 1)


def binary_search_recursive(arr: list, num: int) -> int:
    return __binary_search_rec(arr, num, 0, len(arr) - 1)


def __binary_search_iterate(arr: list, num: int, begin: int, end: int) -> int:
    while begin <= end:
        mid = (begin + end) // 2
        if arr[mid] < num:
            begin = mid + 1
        elif arr[mid] > num:
            end = mid - 1
        else:
            return mid
    return -1


def __binary_search_rec(arr: list, num: int, begin: int, end: int) -> int:
    if begin > end:
        return -1
    mid = (begin + end) // 2
    if arr[mid] == num:
        return mid
    if arr[mid] > num:
        return __binary_search_rec(arr, num, begin, mid - 1)
    else:
        return __binary_search_rec(arr, num, mid + 1, end)
